### PR TITLE
add kernel dialect

### PR DESCRIPTION
**Pull request by jorendumoulin**
start migrating this logic from `util` and formalize them as ops.
Kernel ops can be used to match accelerator templates

### CODE REVIEW

File: tests/filecheck/dialects/kernel/ops.mlir [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// RUN: XDSL_ROUNDTRIP
+// RUN: XDSL_GENERIC_ROUNDTRIP
+
+// CHECK: builtin.module {
+// CHECK-GENERIC: "builtin.module"() ({
+
+%0 = "test.op"() : () -> i32
+// CHECK-NEXT:   %0 = "test.op"() : () -> i32
+// CHECK-NEXT-GENERIC:   %0 = "test.op"() : () -> i32
```
Comment (AntonLydike) [inline tests/filecheck/dialects/kernel/ops.mlir:9]: This should be `CHECK-GENERIC-NEXT:` instead of `CHECK-NEXT-GENERIC`

Comment (jorendumoulin) [reply tests/filecheck/dialects/kernel/ops.mlir:9]: oops! thanks for spotting this!

File: compiler/dialects/kernel.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
+    )
+
+
+@irdl_op_definition
+class MacOp(KernelOp, BinaryOp):
+    name = "kernel.mac"
+    assembly_format = (
+        "$lhs `,` $rhs attr-dict `:`" "type($lhs) `,` type($rhs) `->` type($result)"
+    )
```
Comment (JosseVanDelm) [inline compiler/dialects/kernel.py:46]: Why is this slightly different from the others?
Also, can we not move the assembly format to either ABCs?

File: compiler/dialects/kernel.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+from abc import ABC
+
+from xdsl.dialects.builtin import IntegerType
+from xdsl.ir import Dialect
+from xdsl.irdl import operand_def
+from xdsl.irdl.operations import irdl_op_definition, result_def
+from xdsl.parser import IRDLOperation
+
+
```
Comment (JosseVanDelm) [inline compiler/dialects/kernel.py:9]: Can you add a comment somewhere to state what this dialect's purpose is?

**Review by JosseVanDelm — APPROVED** [review]
I like this! It's nice and short -- I think we can even make it shorter!
Maybe at some point we should add a bit more verification to this, but not necessarily at this point :). Thanks!

File: compiler/dialects/kernel.py (hunk 2 above)
Comment (jorendumoulin) [reply compiler/dialects/kernel.py]: apperently moving them into the abc's breaks xdsl, and it is not recognized. something to do with the @irdl_op_definition magic

**Pull request merged**